### PR TITLE
Improve exit code detection in the slurm driver

### PR DESCRIPTION
**Pull request by verveerpj**
**Issue**
The slurm driver is performing very badly in some cases, when many small jobs were submitted. As it turns out, the use of `scontrol` to extract the exit code of finished jobs may fail, since SLURM does not maintain the information of finished jobs indefinitely in memory for use with `scontrol`. If this happens this leads to many retries of `scontrol`, and the exit code is finally not retrieved. In such a case it is possible to fall back on the `sattr` command which is able to retrieve the information for jobs.

This PR adds `sattr` as a fallback in cases where `scontrol` fails. The `scontrol` command is still preferred and tried first, because depending on the configuration of the SLURM cluster it many not be guaranteed that `sattr` can be used. If that is the case, the code falls back to its old behavior of retrying.

**Approach**
- Add `sattr` command to the slurm options.
- If `scontrol` failure is detected, try `sattr`.
- If `sattr` is not available or fails, fall back on the old retry behavior.

(Screenshot of new behavior in GUI if applicable)

- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'pytest tests/ert/unit_tests -n logical -m "not integration_test"'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

**Comment by berland** [comment]
`_poll_once_by_scontrol() is already a fallback when  `squeue` fails in `poll()`. This will imply a fallback to the fallback for the every-2-second polling. That might be fine though. Intentional?

**Comment by berland** [comment]
There are mocked bash/python scripts for the squeue and scontrol commands. Could you maybe write a mocked sacct command and include tests for this scenario? (that is quite nice to have when developing without access to a real slurm cluster)

**Comment by verveerpj** [comment]
> `_poll_once_by_scontrol() is already a fallback when `squeue`fails in`poll()`. This will imply a fallback to the fallback for the every-2-second polling. That might be fine though. Intentional?

Yes, it seems to be there because `squeue` will not report on jobs that have finished, but we need the exit code. `scontrol` will still report on them for a while, but then it also will fail if too much time has passed. Hence `sacct` as a fallback for the fallback.

**Comment by verveerpj** [comment]
> There are mocked bash/python scripts for the squeue and scontrol commands. Could you maybe write a mocked sacct command and include tests for this scenario? (that is quite nice to have when developing without access to a real slurm cluster)

Yes, will do.

**Comment by verveerpj** [comment]
It looks like a test fails on github because sattr is not set up. I have to look into how to do that, or change the test.

File: tests/ert/unit_tests/scheduler/test_slurm_driver.py [hunk 1]
```diff
@@ -450,3 +459,26 @@ async def finished(iens: int, returncode: int):
     assert not Path(
         "survived"
     ).exists(), "The process children of the job should also have been killed"
+
+
```
Comment (berland) [inline tests/ert/unit_tests/scheduler/test_slurm_driver.py:463]: Would be nice to also test that we can obtain a zero exit code as well, parametrize with `[("true", 0), )("false", 1)]` ?

File: src/ert/scheduler/slurm_driver.py [hunk 2]
```diff
@@ -367,6 +369,18 @@ async def _poll_once_by_scontrol(
         ) and missing_job_id in self._scontrol_cache:
             return self._scontrol_cache[missing_job_id]
 
+        info = await self._run_scontrol(missing_job_id)
+        if info is None:
+            logger.error("scontrol failed, trying sacct")
```
Comment (berland) [inline src/ert/scheduler/slurm_driver.py:374]: This is probably a `logger.warning` since we have a fallback.

File: src/ert/scheduler/slurm_driver.py [hunk 3]
```diff
@@ -384,17 +398,47 @@ async def _poll_once_by_scontrol(
             )
             return None
 
-        info = None
         try:
-            info = _parse_scontrol_output(stdout.decode(errors="ignore"))
+            return _parse_scontrol_output(stdout.decode(errors="ignore"))
         except Exception as err:
             logger.error(
```
Comment (berland) [inline src/ert/scheduler/slurm_driver.py:404]: warning here too then (but I see this is not your change)

File: src/ert/scheduler/slurm_driver.py [hunk 4]
```diff
@@ -384,17 +398,47 @@ async def _poll_once_by_scontrol(
...
+                "-j",
+                str(missing_job_id),
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+            )
+        except FileNotFoundError:
+            logger.error("sacct binary not found")
+            return None
+        stdout, stderr = await process.communicate()
+        if process.returncode:
```
Comment (berland) [inline src/ert/scheduler/slurm_driver.py:427]: `sacct` on my ad-hoc slurm cluster always returns exit code 1 and "Slurm accounting storage is disabled", it seems to be correctly handled here 👍🏻

File: src/ert/scheduler/slurm_driver.py (hunk 2 above)
Comment (verveerpj) [reply src/ert/scheduler/slurm_driver.py]: I agree, I will turn some errors into warnings.

**Review by berland — APPROVED** [review]
Very good! You may want to squash commit 5 into one of the others?

**Comment by verveerpj** [comment]
> Very good! You may want to squash commit 5 into one of the others?

Thanks for the review! Yes, I will squash all into the first.
